### DISCUSSION TITLE
Scheduled event lies in the past

### DISCUSSION THREAD

**Discussion by fabio-r-goncalves** [eclipse/mosaic#326] [Q&A] [answered]
I have built a scenario in a highway where vehicles need to update their speed and driving lane. Everything was going well until I introduced messages. As soon as I did that, I started having the error:

`java.lang.RuntimeException: Scheduled event lies in the past.
        at org.eclipse.mosaic.lib.util.scheduling.DefaultEventScheduler.scheduleEvents(DefaultEventScheduler.java:82)
        at org.eclipse.mosaic.fed.sumo.ambassador.AbstractSumoAmbassador.processTimeAdvanceGrant(AbstractSumoAmbassador.java:1207)
        at org.eclipse.mosaic.rti.api.AbstractFederateAmbassador.advanceTime(AbstractFederateAmbassador.java:110)
        at org.eclipse.mosaic.rti.time.SequentialTimeManagement.runSimulation(SequentialTimeManagement.java:90)
        at org.eclipse.mosaic.starter.MosaicSimulation.runSimulation(MosaicSimulation.java:193)
        at org.eclipse.mosaic.starter.MosaicStarter.execute(MosaicStarter.java:132)
        at org.eclipse.mosaic.starter.MosaicStarter.main(MosaicStarter.java:74)`

Even the lane changes and speed changes are all for future times. Does anyone have any idea what is causing this?
Thanks in advance.

**Comment by fabio-r-goncalves** [comment]
I am almost positive it is caused by a change speedWithInterval called after a message was received. Probably because the simulation time when a message is received is not a round number; it is something like 385032908487, unlike what happens in scheduled events.
How can I solve this?

**Reply by schwepmo** [reply]
It is not uncommon for messages to have these kind of reception numbers, especially when you are using random delays.
From your stacktrace alone it is hard to pinpoint the issue, could you add snippets of your apps, e.g., the parts where you are sending and receiving the messages?

**Reply by fabio-r-goncalves** [reply]
It is a bit hard to send the code, but in summary, what I am doing when the error happens is:

`@Override
    public void onMessageReceived(ReceivedV2xMessage arg0) {
       getOperatingSystem().changeSpeedWithInterval(28, 100000000);
    }`

What I think happens is because the simulation time at the message received time is not a multiple of the simulation time step (100 ms); when I try to change speed, it will create an event to a little after the next timestep. So, it will only handle it after two timesteps when, in reality, the event is in the past.

I think I was able to fix that. So, now, if the simulation time is not a simulation time step, I calculate the needed offset for it to be and thus force the changeSpeedWithInterval to be in a multiple.

**Reply by schwepmo** [reply, marked as answer]
Hey, I was able to reproduce this issue and also think I found the error source and the issue was exactly what you've described. Looks like we had a little bit of a regression during one of our refactors.

https://github.com/eclipse/mosaic/blob/7edbd224193dbe201a96846ebbc5d34677af0725/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L873
If you change the above line in the `AbstractSumoAmbassador` as follows, the times should be properly adjusted to the next simulation step in SUMO. Alternatively, you just wait a couple of days and merge the main-branch into yours.
```java
long adjustedTime = adjustToSumoTimeStep(changeSpeedTimestep, sumoConfig.updateInterval * TIME.MILLI_SECOND); 
```

Sorry for that inconvenience and thanks for pointing it out!
